core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1019 @@
+      // for a fresh start.
+      if (detectionTimerHandle != null) {
+        detectionTimerHandle.cancel();
+        trackerMap.resetCallCounters();
+      }
+
+      detectionTimerHandle = syncContext.scheduleWithFixedDelay(new DetectionTimer(config),
+          initialDelayNanos, SECONDS.toNanos(config.intervalSecs), NANOSECONDS, timeService);
+    } else if (detectionTimerHandle != null) {
+      // Outlier detection is not configured, but we have a lingering timer. Let's cancel it and
+      // uneject any addresses we may have ejected.
+      detectionTimerHandle.cancel();
+      detectionTimerStartNanos = null;
+      trackerMap.cancelTracking();
+    }
+
+    switchLb.handleResolvedAddresses(resolvedAddresses);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    switchLb.handleNameResolutionError(error);
+  }
+
+  @Override
+  public void shutdown() {
+    switchLb.shutdown();
+  }
+
+  /**
+   * This timer will be invoked periodically, according to configuration, and it will look for any
+   * outlier subchannels.
+   */
+  class DetectionTimer implements Runnable {
+
+    OutlierDetectionLoadBalancerConfig config;
+
+    DetectionTimer(OutlierDetectionLoadBalancerConfig config) {
+      this.config = config;
+    }
+
+    @Override
+    public void run() {
+      detectionTimerStartNanos = timeProvider.currentTimeNanos();
+
+      trackerMap.swapCounters();
+
+      for (OutlierEjectionAlgorithm algo : OutlierEjectionAlgorithm.forConfig(config)) {
+        algo.ejectOutliers(trackerMap, detectionTimerStartNanos);
+      }
+
+      trackerMap.maybeUnejectOutliers(detectionTimerStartNanos);
+    }
+  }
+
+  /**
+   * This child helper wraps the provided helper so that is can hand out wrapped {@link
+   * OutlierDetectionSubchannel}s and manage the address info map.
+   */
+  class ChildHelper extends ForwardingLoadBalancerHelper {
+
+    private Helper delegate;
+
+    ChildHelper(Helper delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    protected Helper delegate() {
+      return delegate;
+    }
+
+    @Override
+    public Subchannel createSubchannel(CreateSubchannelArgs args) {
+      // Subchannels are wrapped so that we can monitor call results and to trigger failures when
+      // we decide to eject the subchannel.
+      OutlierDetectionSubchannel subchannel = new OutlierDetectionSubchannel(
+          delegate.createSubchannel(args));
+
+      // If the subchannel is associated with a single address that is also already in the map
+      // the subchannel will be added to the map and be included in outlier detection.
+      List<EquivalentAddressGroup> allAddresses = subchannel.getAllAddresses();
+      if (allAddresses.size() == 1 && trackerMap.containsKey(allAddresses.get(0))) {
+        AddressTracker eagInfo = trackerMap.get(allAddresses.get(0));
+        subchannel.setEquivalentAddressGroupInfo(eagInfo);
```

Comment:
This needs to be called on `subchannel.updateAddresses()` as well.

Reply:
the bidirectional association between the map and the trackers is now managed in the add/removeTracker() methods so things should always stay in sync.